### PR TITLE
[h264e] Add EncTools support in AVC Encoder

### PR DESCRIPTION
**Pull request by takulako**

### CODE REVIEW

**Comment by takulako** [comment]
Pull request https://github.com/Intel-Media-SDK/MediaSDK/pull/2271/ was rebased, code duplication minimized in accordance with comments, submitted as a new request

**Comment by onabiull** [comment]
Please address clang  10 failures:
```
/docker/worker/mediasdk-clang-10_0/build_dir/repos/MediaSDK/_studio/enctools/include/mfx_enctools_brc_defs.h:832:15: error: function 'GetOffsetAPQ' has internal linkage but is not defined [-Werror,-Wundefined-internal]
/docker/worker/mediasdk-clang-10_0/build_dir/repos/MediaSDK/_studio/enctools/include/mfx_enctools_brc_defs.h:832:15: error: function 'GetOffsetAPQ' has internal linkage but is not defined [-Werror,-Wundefined-internal]
```

File: _studio/shared/include/mfx_utils.h [hunk 1]
```diff
@@ -37,33 +37,13 @@
...
-#define MFX_CHECK(EXPR, ERR)    { if (!(EXPR)) MFX_RETURN(ERR); }
 
 #define MFX_CHECK_NO_RET(EXPR, STS, ERR){ if (!(EXPR)) { std::ignore = MFX_STS_TRACE(ERR); STS = ERR; } }
 
-#define MFX_CHECK_STS(sts)              MFX_CHECK(MFX_SUCCEEDED(sts), sts)
 #define MFX_SAFE_CALL(FUNC)             { mfxStatus _sts = FUNC; MFX_CHECK_STS(_sts); }
-#define MFX_CHECK_NULL_PTR1(pointer)    MFX_CHECK(pointer, MFX_ERR_NULL_PTR)
-#define MFX_CHECK_NULL_PTR2(p1, p2)     { MFX_CHECK(p1, MFX_ERR_NULL_PTR); MFX_CHECK(p2, MFX_ERR_NULL_PTR); }
-#define MFX_CHECK_NULL_PTR3(p1, p2, p3) { MFX_CHECK(p1, MFX_ERR_NULL_PTR); MFX_CHECK(p2, MFX_ERR_NULL_PTR); MFX_CHECK(p3, MFX_ERR_NULL_PTR); }
+
```
Comment (dvrogozh) [inline _studio/shared/include/mfx_utils.h:46]: I still don't see the point to move only part of these definitions to mfx_utils_defs.h. I believe you just need to move all of these defs to this new file regardless whether you use them in a new code or not.

**Review by dvrogozh — CHANGES_REQUESTED** [review]

**Review by splotnikv — APPROVED** [review]

File: _studio/shared/include/mfx_utils.h (hunk 1 above)
Comment (takulako) [reply _studio/shared/include/mfx_utils.h]: done

**Review by dvrogozh — CHANGES_REQUESTED** [review]
Please, do couple of formal things before this can be merged:
1. Cleanup history. Likely there should be single commit introducing the feature
2. Update https://github.com/Intel-Media-SDK/MediaSDK/blob/master/CODEOWNERS with people who should review changes to newly added files.

**Review by dvrogozh — CHANGES_REQUESTED** [review]
Please, clean up commits history. Since this is a new feature there likely should be single commit introducing it. Provide comprehensive commit message:
* How to enable the feature on a build level (it seems disabled by default)
* How to enable on apps level, give samples command lines examples if applicable

**Comment by dvrogozh** [comment]
And this PR effectively does not build:
`/docker/worker/mediasdk-clang-10_0/build_dir/repos/MediaSDK/_studio/enctools/include/mfx_enctools_brc_defs.h:832:15: error: function 'GetOffsetAPQ' has internal linkage but is not defined [-Werror,-Wundefined-internal]
`

**Review by dvrogozh — COMMENTED** [review]
Resolve conflicts, please.

File: builder/ConfTargets.cmake [hunk 2]
```diff
@@ -68,6 +68,8 @@ else()
   set ( MFX_1_34_OPTIONS_ALLOWED OFF )
 endif()
 
+option( MFX_ENABLE_AENC "Enabled AENC extension?" ON)
```
Comment (dvrogozh) [inline builder/ConfTargets.cmake:71]: No! You can _not_ enable enctools by default till you rely on closed source binary (libaenc.a).

File: _studio/enctools/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    ${CMAKE_CURRENT_SOURCE_DIR}/src/mfx_enctools_common.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/src/mfx_enctools_aenc.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/src/mfx_enctools_utils.cpp
+    )
+set( sources.plus "" )
+
+set( defs "${API_FLAGS} ${WARNING_FLAGS}" )
+
+if (MFX_ENABLE_AENC)
+  find_library( AENC_LIBRARY libaenc.a PATHS ${CMAKE_HOME_DIRECTORY}/contrib/aenc/ )
```
Comment (dvrogozh) [inline _studio/enctools/CMakeLists.txt:21]: I don't quite get why you need so many enable defines:
* MFX_ENABLE_AENC
* MFX_ENABLE_ENCTOOLS
* MFX_ENCTOOLS_ON

It seems that at least one of them is not needed. Can we have the following?
* MFX_ENABLE_ENCTOOLS=ON|OFF, OFF by default
* MFX_ENABLE_AENC=ON|OFF, OFF by default

File: samples/sample_common/include/brc_routines.h [hunk 4]
```diff
@@ -16,8 +16,11 @@ This sample was distributed or derived from the Intel's Media Samples package.
 The original version of this sample may be obtained from https://software.intel.com/en-us/intel-media-server-studio
 or https://software.intel.com/en-us/media-client-solutions-support.
 \**********************************************************************************/
+
+#ifndef __PIPELINE_ENCODE_BRC_H__
```
Comment (dvrogozh) [inline samples/sample_common/include/brc_routines.h:20]: emm... Why did you move that below?

File: samples/sample_common/include/brc_routines.h [hunk 5]
```diff
@@ -52,12 +55,14 @@ NalHrdConformance | VuiNalHrdParameters   |  Result
     on (or default)      on (or default)    => MFX_BRC_HRD_STRONG
 --------------------------------------------------------------
 */
+#if !defined(__MFXENCTOOLS_INT_H__) || (MFX_VERSION < MFX_VERSION_NEXT)
```
Comment (dvrogozh) [inline samples/sample_common/include/brc_routines.h:58]: This does not make sense to me: that was already openly defined...

File: samples/sample_common/src/brc_routines.cpp [hunk 6]
```diff
@@ -19,8 +19,8 @@ or https://software.intel.com/en-us/media-client-solutions-support.
 
 
 #include "brc_routines.h"
-#include "math.h"
-#include "mfxdefs.h"
+
+#include <math.h>
```
Comment (dvrogozh) [inline samples/sample_common/src/brc_routines.cpp:23]: I don't see that you give any sample example on how to use enctools...

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: mfxconfig.h.in [hunk 7]
```diff
@@ -44,3 +44,5 @@
 #cmakedefine MFX_ENABLE_MJPEG_VIDEO_ENCODE
 #cmakedefine MFX_ENABLE_VC1_VIDEO_DECODE
 #cmakedefine MFX_ENABLE_AV1_VIDEO_DECODE
+
+#cmakedefine MFX_ENABLE_ENCTOOLS
```
Comment (dvrogozh) [inline mfxconfig.h.in:48]: This one is not seems hanging in the air. You only expose cmake option for MFX_ENABLE_AENC, but specify MFX_ENABLE_ENCTOOLS here and don't have MFC_ENABLE_AENC at all. This confuses a lot!

Let's just stop all that and have single MFX_ENABLE_ENCTOOLS for everything?

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: samples/sample_common/src/brc_routines.cpp (hunk 6 above)
Comment (dvrogozh) [reply samples/sample_common/src/brc_routines.cpp:23]: done in the recent commit msg

File: _studio/enctools/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    add_library(aenc STATIC IMPORTED GLOBAL)
+    set_target_properties(aenc PROPERTIES IMPORTED_LOCATION "${CMAKE_HOME_DIRECTORY}/contrib/aenc/libaenc.a")
+    target_include_directories(aenc INTERFACE ${CMAKE_CURRENT_SOURCE_DIR}/include)
+  else()
+    message(STATUS "aenc library was not found (optional)! mfxlib will be built without aenc")
+    set(MFX_ENABLE_AENC OFF)
+  endif()
+endif()
+
+if (MFX_ENABLE_AENC)
```
Comment (dvrogozh) [inline _studio/enctools/CMakeLists.txt:34]: you don't need this block at all. Actually, add `#cmakedefine MFX_ENABLE_AENC` in the mfxconfig.h.in and what you are doing will be done automatically and in a more clear way.

File: mfxconfig.h.in (hunk 7 above)
Comment (dvrogozh) [inline mfxconfig.h.in:48]: Add:
```
#cmakedefine MFX_ENABLE_AENC
```

File: samples/sample_common/include/brc_routines.h [hunk 9]
```diff
@@ -52,12 +55,14 @@ NalHrdConformance | VuiNalHrdParameters   |  Result
     on (or default)      on (or default)    => MFX_BRC_HRD_STRONG
 --------------------------------------------------------------
 */
+#if !defined(__MFXENCTOOLS_INT_H__)
```
Comment (dvrogozh) [inline samples/sample_common/include/brc_routines.h:58]: This is a wrong change! `_studio/shared/include/mfxenctools-int.h` is mediasdk internal file, not an API header! You just can't refer to it in samples.

Effectively, I don't see that any of the changes you've mode to samples bring any value. They should completely be dropped.

File: tools/tracer/CMakeLists.txt [hunk 10]
```diff
@@ -70,7 +70,7 @@ get_mfx_version(mfx_version_major mfx_version_minor)
 set_target_properties(mfx-tracer PROPERTIES   VERSION ${mfx_version_major}.${mfx_version_minor})
 set_target_properties(mfx-tracer PROPERTIES SOVERSION ${mfx_version_major})
 
-target_link_libraries( mfx-tracer ${CMAKE_DL_LIBS})
+target_link_libraries( mfx-tracer PRIVATE ${CMAKE_DL_LIBS})
```
Comment (dvrogozh) [inline tools/tracer/CMakeLists.txt:73]: This is also unrelated change. Focus on enctools PR, please. Drop this change.

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: _studio/enctools/CMakeLists.txt (hunk 8 above)
Comment (takulako) [reply _studio/enctools/CMakeLists.txt:34]: What if MFX_ENABLE_AENC=ON is set at build but libaenc.a is not found? We need to have  MFX_ENABLE_AENC undefined  in the code in this case. In other words, we need to have a define defined/undefined depending on the presence of libanenc.a

Comment (takulako) [reply _studio/enctools/CMakeLists.txt:34]: Or we just don't need to consider such a situation?

File: _studio/enctools/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    ${CMAKE_CURRENT_SOURCE_DIR}/src/mfx_enctools_brc.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/src/mfx_enctools_common.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/src/mfx_enctools_aenc.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/src/mfx_enctools_utils.cpp
+    )
+set( sources.plus "" )
+
+set( defs "${API_FLAGS} ${WARNING_FLAGS}" )
+
+if (MFX_ENABLE_AENC)
```
Comment (dvrogozh) [inline _studio/enctools/CMakeLists.txt:20]: I think we better move this place to builder/FindPackages.cmake where we have single top level place to detect availability of our dependencies.

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: _studio/enctools/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+  find_library( AENC_LIBRARY libaenc.a PATHS ${CMAKE_HOME_DIRECTORY}/contrib/aenc/ )
+  if(NOT ${AENC_LIBRARY} MATCHES NOTFOUND)
+    message(STATUS "aenc library was found here ${AENC_LIBRARY}")
+
+    add_library(aenc STATIC IMPORTED GLOBAL)
+    set_target_properties(aenc PROPERTIES IMPORTED_LOCATION "${CMAKE_HOME_DIRECTORY}/contrib/aenc/libaenc.a")
+    target_include_directories(aenc INTERFACE ${CMAKE_CURRENT_SOURCE_DIR}/include)
+  else()
+    message(STATUS "aenc library was not found (optional)! mfxlib will be built without aenc")
```
Comment (dvrogozh) [inline _studio/enctools/CMakeLists.txt:29]: Better to stop immediately with error. User requested this explicitly - no need to proceed if not found.

File: _studio/enctools/CMakeLists.txt (hunk 8 above)
Comment (dvrogozh) [reply _studio/enctools/CMakeLists.txt:34]: We just need to stop with FATAL_ERROR message on cmake configuration stage.

**Comment by dvrogozh** [comment]
I am ok with the change, but you need to fix build/tests when enctools are enabled: see https://github.com/Intel-Media-SDK/MediaSDK/pull/2434 for errors.

File: _studio/enctools/CMakeLists.txt (hunk 11 above)
Comment (onabiull) [reply _studio/enctools/CMakeLists.txt:22]: It lead to chicken-egg like problem :) I suggest to keep here until bigger refactoring.

**Review by onabiull — APPROVED** [review]

**Review by dvrogozh — APPROVED** [review]
+1 considering that build issues for enabled enctools were fixed (#2434).

**Review by DenWolf — APPROVED** [review]

**Pull request merged**